vnext/Microsoft.ReactNative/Utils/TextHitTestUtils.cpp

```diff
@@ -0,0 +1,328 @@
+ * ┌─────> X
+ * │ (x,y) ┌──────┐
+ * │ ┌─────┘      │
+ * │ │            │
+ * ↓ │   ┌─┐   ┌──┘
+ * Y └─>─┘<└─<─┘
+ *            ↑
+ */
+static BidirectionalTextBoundary FindEndRTLBoundary(
+    const xaml::Documents::TextPointer &start,
+    const xaml::Documents::TextPointer &end,
+    const winrt::Rect &target) {
+  auto textPointer = start;
+  auto L = start.Offset();
+  auto R = end.Offset();
+  while (L <= R) {
+    const auto m = /* floor */ (L + R) / 2;
+    const auto relativeOffset = m - textPointer.Offset();
+    textPointer = textPointer.GetPositionAtOffset(relativeOffset, xaml::Documents::LogicalDirection::Forward);
+    const auto rect = textPointer.GetCharacterRect(xaml::Documents::LogicalDirection::Forward);
+    // If median character is on different line, boundary is after median
+    if (rect.Y != target.Y) {
+      L = m + 1;
+    } else if (m < R) {
+      const auto nextPointer = textPointer.GetPositionAtOffset(1, xaml::Documents::LogicalDirection::Forward);
+      const auto nextRect = nextPointer.GetCharacterRect(xaml::Documents::LogicalDirection::Forward);
+      if (nextRect.X < target.X) {
+        // If next character is left of end, boundary is after median
+        L = m + 1;
+      } else if (rect.X >= target.X) {
+        // If median character is right of end, boundary is before median
+        R = m - 1;
+      } else {
+        // Otherwise, we've found the boundary where the next character is the
+        // first RTL character and the median is the last LTR character
+        BidirectionalTextBoundary result;
+        result.rtlPointer = nextPointer;
+        result.rtlRect = nextRect;
+        result.ltrPointer = textPointer;
+        result.ltrRect = rect;
+        return result;
+      }
+    } else {
+      break;
+    }
+  }
+
+  // No LTR characters on last line, L points to first character on last line
+  auto lastRTL = end;
+  auto lastRTLRect = target;
+  if (L < end.Offset()) {
+    const auto relativeOffset = L - end.Offset();
+    lastRTL = end.GetPositionAtOffset(relativeOffset, xaml::Documents::LogicalDirection::Forward);
+    lastRTLRect = lastRTL.GetCharacterRect(xaml::Documents::LogicalDirection::Forward);
+  }
+
+  BidirectionalTextBoundary result;
+  result.rtlPointer = lastRTL;
+  result.rtlRect = lastRTLRect;
+  return result;
+}
+
+/**
+ * Determine if the next character in the given direction on the same line is
+ * RTL. If the next character is not on the same line, assume LTR. This may be
+ * The case if the run has only one character on the target line.
+ */
+static bool IsRTL(const xaml::Documents::TextPointer &textPointer, const winrt::Rect &rect, int direction) {
+  auto currentPointer = textPointer.GetPositionAtOffset(direction, xaml::Documents::LogicalDirection::Forward);
+  while (currentPointer != nullptr) {
+    const auto currentRect = currentPointer.GetCharacterRect(xaml::Documents::LogicalDirection::Forward);
+    if (currentRect.Y != rect.Y) {
+      return false;
```

Comment:
more FP equality comparisons here and below give me pause :D 
Ideally we make sure that all these comparisons are done within the same machine epsilon, so maybe extract it to a common helper?

Reply:
Commented above, we can switch to epsilon, but not sure it's necessary.